README.md
# globalFetch

[![Version](https://img.shields.io/npm/v/js-global-fetch.svg)](https://www.npmjs.org/package/js-global-fetch) [![Extension](https://www.google.com/images/icons/product/chrome_web_store-32.png)](https://alex2844.github.io/js-global-fetch/index.html?extension)

## Universal Fetch API, FormData, AbortController, CorsProxy for Nodejs, Browsers and Google Apps Script

#### Install
###### nodejs
``` bash
npm install js-global-fetch
```
###### cdn
``` html
<script src="https://alex2844.github.io/js-global-fetch/dist/polyfill.js"></script>
```
###### Google Apps Script (https://github.com/alex2844/js-global-fetch/blob/master/GAS.md)
###### extension
- chrome://extensions/
- Dev mode on
- 'Load unpacked extension' button and select the unzipped folder for your extension to install it.

#### Usage
###### Get json
``` javascript
const { fetch } = require('js-global-fetch');
fetch('https://httpbin.org/get', {
	headers: {
		'Cors-Cache': 60*60*24*1000,
		'Cors-Referer': 'https://alex2844.github.io/js-global-fetch/',
		'Cors-User-Agent': 'globalFetch'
	}
})
.then(res => res.json())
.then(body => console.log(body));
```
###### CorsProxy as a plugin for Fastify
``` javascript
const
	Fastify = require('fastify'),
	{ CorsProxy } = require('js-global-fetch');
(Fastify()
	.register(CorsProxy, { prefix: '/proxy' }) // http://localhost:3000/proxy/
	.get('/', async (req, rep) => 'Index page') // http://localhost:3000/
	.listen(3000, () => console.log('Server start'))
);
```
###### Get iframe content
``` html
<iframe src="https://httpbin.org/get" frameborder="0" allowfullscreen="true" width="500" height="300" loading="lazy"></iframe>
<script>
	$('iframe').addEventListener('load', event => {
		// console.log(event.target.contentWindow.document.body.innerHTML); // Blocked a frame with origin "http://localhost:8080" from accessing a cross-origin frame.
		new Promise(res => {
			let onmessage = e_ => {
				if (e_.data.corsProxy && (event.timeStamp == e_.data.timeStamp)) {
					window.removeEventListener('message', onmessage, false);
					return res(e_.data._corsProxy_);
				}
			}
			event.target.contentWindow.postMessage({
				corsProxy: {
					eval: (() => document.body.innerHTML).toString()
				},
				timeStamp: event.timeStamp
			}, '*');
			window.addEventListener('message', onmessage, false);
		}).then(html => {
			console.log(JSON.parse(html.replace(/^(.*?)>/, '').replace(/<(.*?)$/, '')));
		});
	}, { once: true });
</script>
```
###### Open "Youtube on TV"
https://www.youtube.com/tv?corsProxy={%22user-agent%22:%22Mozilla/%20Chrome/%20Mobile%20Safari/;%20SMART-TV%22}

#### Run cli
``` bash
PORT=3000 globalFetch
```

#### Demo and tests
https://alex2844.github.io/js-global-fetch/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.  
Added for new features. (Добавлено для новых функций.)  
Changed for changes in existing functionality. (Изменено для изменений в существующей функциональности.)  
Deprecated for soon-to-be removed features. (Устарело для функций, которые скоро будут удалены.)  
Removed for now removed features. (Удалено для удалённых функций.)  
Fixed for any bug fixes. (Исправлено для любых исправлений.)  
Security in case of vulnerabilities. (Безопасность в случае уязвимостей.)

## [Unreleased]

## [1.6.6] - 2021-10-15
### Fixed
- add background.html

## [1.6.5] - 2021-10-14
### Added
- background.html

## [1.6.4] - 2021-10-12
### Added
- public background.menu && background.config

## [1.6.3] - 2021-08-28
### Added
- iframe eval support promise

## [1.6.2] - 2021-08-28
### Added
- Get iframe webRequests

## [1.6.1] - 2021-04-23
### Fixed
- JS detect ua
- JS detect referer
- only contentType == text/html

## [1.5.1] - 2021-01-21
### Added
- Get iframe content
### Fixed
- Check module parent

## [1.5.0] - 2020-10-25
### Added
- Support Google Apps Script

## [1.2.1] - 2020-04-24
### Added
- Cors-Cache (worker)
- Add chrome extension

## [1.0.0] - 2020-03-18
### Added
- Init

[Unreleased]: ../../compare/v1.5.1...HEAD
[1.5.1]: ../../releases/tag/v1.5.1
[1.5.0]: ../../releases/tag/v1.5.0
[1.2.1]: ../../releases/tag/v1.2.1
[1.0.0]: ../../releases/tag/v1.0.0


test.js
var self = ((typeof(window) == 'object') ? window : global);
var worker;
var fn = {
	fetch: null,
	Request: null,
	AbortController: null
};
var tests = {
	corsWorker: function() {
		return fetch(worker+'https://api.qwedl.com/ip.php?c=1&no_cors', {
			headers: {
				'Content-type': 'application/json',
				'Cors-Referer': null,
				'Cors-User-Agent': 'corsWorker'
			}
		}).then(function(res) {
			return res.json().then(function(body) {
				return ((!body.headers.Referer && (body.headers['User-Agent'] == 'corsWorker') && (res.headers.get('x-test') == 'Test 2')) && 'ok');
			});
		}).catch(function(err) {
			return 'Error: '+err.message;
		});
	},
	corsUpload: function(e) {
		var file,
			data = new FormData();
		if ('document' in self)
			try {
				file = new File([ JSON.stringify({ name: 'js-global-fetch' }) ], 'package.json', { type:'application/json' });
			} catch (e) {
				file = new Blob([ JSON.stringify({ name: 'js-global-fetch' }) ], { type:'application/json' });
				file.lastModified = file.lastModifiedDate = new Date();
				file.name = 'package.json';
			}
		else
			file = require('fs').createReadStream('package.json');
		data.append('file', file);
		return fetch(worker+'https://api.qwedl.com/ip.php?format=json&no_cors', {
			method: 'POST',
			body: data
		}).then(function(res) {
			return res.json().then(function(body) {
				return ((
					body.data.files.file.size && (body.data.files.file.name == 'package.json') &&
					(body.data.files.file.type == 'application/json') && (body.data.files.file.data.name == 'js-global-fetch')
				) && 'ok');
			});
		}).catch(function(err) {
			console.info(err);
			return 'Error: '+err.message;
		});
	},
	corsForm: function(e) {
		return fetch(worker+'https://api.qwedl.com/ip.php?format=json&no_cors', {
			body: 'asd1=qwe&asd2=qwe',
			method: 'POST',
			headers: { 'content-type': 'application/x-www-form-urlencoded' }
		}).then(function(res) {
			return res.json().then(function(body) {
				return (((body.data.post.asd1 == 'qwe') && (body.data.post.asd2 == 'qwe')) && 'ok');
			});
		}).catch(function(err) {
			return 'Error: '+err.message;
		});
	},
	corsOk: function() {
		return (!worker ? Promise.resolve('worker not found') : fetch(worker).then(function(res) {
			return res.text();
		}).then(function(body) {
			return ((body == worker+'<url>') && 'ok');
		}).catch(function(err) {
			return 'Error: '+err.message;
		}));
	},
	corsErr: function() {
		return (!worker ? Promise.resolve('worker not found') : fetch(worker+'/iscorsneeded').then(function(res) {
			return res.text();
		}).then(function(body) {
			return ((body == (('document' in self) ? 'yes' : 'no')) && 'ok');
		}).catch(function(err) {
			return (((err.message == 'Failed to fetch') || (err.message.indexOf('is not allowed by Access-Control-Allow-Origin.') > -1)) ? 'ok' : 'Error: '+err.message);
		}));
	},
	corsVideo: function() {
		return fetch(worker+'https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4', {
			method: 'HEAD'
		}).then(function(res) {
			return (((res.headers.get('content-length') == '158008374') && (res.headers.get('content-type') == 'video/mp4')) && 'ok');
		}).catch(function(err) {
			return 'Error: '+err.message;
		});
	},
	timeout: function() {
		var acs = new AbortController();
		setTimeout(function() {
			acs.abort();
		}, 10);
		return fetch('https://api.qwedl.com/ip.php', {
			signal: acs.signal
		}).then(function(res) {
			return res.text();
		}).catch(function(err) {
			return ((err.name == 'AbortError') ? 'ok' : 'Error: '+err.message);
		});
	}
};
var count = Object.keys(tests).length;
function log(obj) {
	if ('document' in self) {
		if (!count)
			document.querySelector('.content header button').hidden = true;
		var el = document.querySelector('code');
		el.innerHTML = JSON.stringify(obj, null, '\t');
		el.classList.remove('prettyprinted');
		if ('PR' in window)
			PR.prettyPrint();
	}else{
		console.clear();
		console.log(obj);
	}
}
Object.keys(fn).forEach(function(f) {
	if (!!self[f])
		fn[f] = 'native';
});
log({ fn });
if ('document' in self) {
	if (!Object.keys(fn).filter(function(f) {
		return !fn[f];
	}).length)
		load();
	else{
		var polyfill = document.createElement('script');
		polyfill.src = './dist/polyfill.js';
		polyfill.onload = load;
		document.body.appendChild(polyfill);
	}
}else{
	var { fetch, Request, FormData, AbortController } = polyfill = require('./');
	load();
}
function load() {
	Object.keys(fn).filter(function(f) {
		return !fn[f];
	}).forEach(function(f) {
		if (!!self[f] || !!polyfill[f])
			fn[f] = 'polyfill';
	});
	log({ fn });
	if ('document' in self) {
		if ('worker' in sessionStorage)
			worker = sessionStorage.getItem('worker');
		else if ((worker = prompt('Worker url')) != null)
			sessionStorage.setItem('worker', worker);
		if (worker === null)
			worker = '';
		tests_();
	}else{
		if (worker = process.env.WORKER)
			tests_();
		else{
			var rl = require('readline').createInterface({
				input: process.stdin,
				output: process.stdout
			});
			rl.question('Worker url: ', function(worker_) {
				worker = worker_;
				rl.close();
				tests_();
			});
		}
	}
}
function tests_() {
	var _tests = {};
	for (var k in tests) {
		(function(k) {
			tests[k]().then(function(res) {
				--count;
				_tests[k] = res;
				log({
					worker: (worker || (('document' in self) && document.documentElement.dataset.corsProxy && JSON.parse(document.documentElement.dataset.corsProxy))),
					fn, _tests
				});
			});
		})(k);
	}
}


gulpfile.js
const
	{ fetch } = require('./'),
	fs = require('fs'),
	path = require('path'),
	gulp = require('gulp'),
	plumber = require('gulp-plumber'),
	uglify = require('gulp-uglify');

gulp.task('AbortController', done => {
	gulp.src(path.join(__dirname, 'node_modules', 'abort-controller', 'dist', 'abort-controller.js'))
		.pipe(gulp.dest('dist'))
		.on('end', () => done())
});
gulp.task('polyfill', done => {
	let fn = path.join(__dirname, 'dist', 'polyfill.js');
	fetch('https://polyfill.io/v3/polyfill.js?features=AbortController', {
		headers: { 'user-agent': 'Mozilla/1.22 (compatible; MSIE 10.0; Windows 3.1)' }
	}).then(res => res.text()).then(body => {
		fs.writeFile(fn, body.replace('function fetch', (function fetch_gas(input, init) {
			return Promise.resolve().then(function() {
				var req = UrlFetchApp.fetch(input, init),
					code = req.getResponseCode();
				return new Response(req.getContentText(), {
					status: code,
					statusText: (code === 200 ? 'OK' : ''),
					headers: req.getHeaders(),
					url: input
				});
			})
		}).toString()+'\n\n  function fetch_xhr').replace('fetch.polyfill', (function fetch(input, init) {
			return ((typeof(XMLHttpRequest) == 'function') ? fetch_xhr(input, init) : (
				(typeof(UrlFetchApp) == 'object') ? fetch_gas(input, init) : Promise.reject(new Error('Hmmm...'))
			));
		}).toString()+'\n\n	fetch.polyfill'), err => {
			if (err)
				return console.log(err);
			gulp.src(fn)
				.pipe(plumber())
				.pipe(uglify())
				.pipe(gulp.dest('dist'))
				.on('end', () => done())
		});
	}).catch(() => done());
});
gulp.task('default', gulp.series('AbortController', 'polyfill'));


globalFetch.js
#!/usr/bin/env node

// process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const
	NodeFetch = require('node-fetch'),
	FormData = require('form-data');

let AbortController = {};
try {
	AbortController = require('./dist/abort-controller');
} catch (e) {
	console.error('Exec: npm run build');
}

module.exports = {
	CorsProxy: async (fastify, opts) => (fastify
		.addContentTypeParser('*', (req, done) => done(null, req))
		.all('/*', async (req, rep) => {
			let res = {
				status: 200,
				headers: {},
				body: ''
			};
			try {
				let query = (new URLSearchParams(req.query)).toString(),
					url = req.params['*']+(query ? '?'+query : '');
				if (!url) {
					url = ((req.headers['x-forwarded-proto'] && req.headers['x-forwarded-proto'].startsWith('https')) ? 'https:' : 'http:')+'//'+req.hostname+req.raw.url;
					res.body = url.replace(/\/$/, '')+'/<url>';
				}else if (url == '/iscorsneeded')
					return (req.headers['sec-fetch-dest'] ? 'yes' : 'no');
				else{
					if (req.raw.method != 'OPTIONS') {
						let headers = {};
						for (let k in req.headers) {
							if (!k.match(/^host/i) && !k.match(/^origin/i) && !k.match(/^cf-/i) && !k.match(/^x-forw/i) && !k.match(/^cors-/i))
								headers[k] = req.headers[k];
						}
						for (let k in req.headers) {
							let k_;
							if (k.match(/^cors-/i) && (k_ = k.replace(/^cors-/i, ''))) {
								if (req.headers[k] && (req.headers[k] != 'null'))
									headers[k_] = req.headers[k];
								else
									delete headers[k_];
							}
						}
						let get = await NodeFetch(url, {
							method: req.raw.method,
							headers: headers,
							redirect: 'follow',
							body: req.body
						});
						headers = {};
						for (let pair of get.headers.entries()) {
							if ([ 'keep-alive', 'transfer-encoding', 'content-encoding' ].indexOf(pair[0]) == -1)
								headers[pair[0]] = pair[1];
						}
						res.status = get.status;
						res.headers = headers;
						res.body = get.body;
					}
				}
				if (req.headers['origin']) {
					res.headers['Access-Control-Allow-Origin'] = req.headers['origin'];
					res.headers['Access-Control-Allow-Credentials'] = 'true';
					if (req.headers['access-control-request-headers'])
						res.headers['Access-Control-Allow-Headers'] = req.headers['access-control-request-headers'];
					res.headers['Access-Control-Expose-Headers'] = '*';
				}
			} catch (e) {
				res.status = 500;
				res.body = e.message;
			}
			rep.status(res.status || 200).headers(res.headers || {}).send(res.body || '');
		})
	),
	fetch: (url, options) => {
		if (/^\/\//.test(url))
			url = 'http:'+url;
		if (options && options.headers)
			for (let k in options.headers) {
				let k_;
				if (k.match(/^cors-/i) && (k_ = k.replace(/^cors-/i, ''))) {
					if (options.headers[k] && (options.headers[k] != 'null'))
						options.headers[k_] = options.headers[k];
					delete options.headers[k];
				}
			}
		return (NodeFetch.default || NodeFetch).call(this, url, options);
	},
	Headers: NodeFetch.Headers,
	Request: NodeFetch.Request,
	Response: NodeFetch.Response,
	FormData,
	AbortController
};
if (module.parent === null)
	module.exports.CorsProxy(require('fastify')()).then(fastify => {
		fastify.listen((process.env.PORT || 0), () => console.log('Server started http://localhost:'+fastify.server.address().port));
	});


extension/content.js
let { name, version } = chrome.runtime.getManifest();
if (document.contentType == 'text/html') {
	if (location.search) {
		let corsProxy = (Object.fromEntries(new URLSearchParams(location.search)).corsProxy || '').replace(/"([^"]+)":/g, (s, k) => ('"'+k.toLowerCase()+'":'));
		if (corsProxy) {
			let code = '',
				corsProxy_ = JSON.parse(corsProxy);
			if (corsProxy_['user-agent'] && (corsProxy_['user-agent'] != navigator.userAgent)) {
				code +=  '('+function(corsProxy) {
					Object.defineProperty(navigator, 'userAgent', {
						get: () => corsProxy['user-agent']
					});
				}+')('+corsProxy+');';
			}
			if (corsProxy_['referer'] && (corsProxy_['referer'] != document.referrer)) {
				code +=  '('+function(corsProxy) {
					Object.defineProperty(document, 'referrer', {
						get: () => corsProxy['referer']
					});
				}+')('+corsProxy+');';
			}
			if (code) {
				let s = document.createElement('script');
				s.textContent = code;
				document.documentElement.appendChild(s);
				s.remove();
			}
		}
	}
	chrome.storage.local.get({
		enabled: true
	}, config => {
		if (config.enabled) {
			document.documentElement.setAttribute('data-cors-proxy', JSON.stringify({
				id: chrome.runtime.id,
				name, version, config
			}));
			let webRequests;
			chrome.runtime.onMessage.addListener(res => {
				if (res.webRequest && !!webRequests)
					webRequests.push(res.webRequest);
			});
			window.addEventListener('message', async event => {
				if (event.data.corsProxy && !event.data.runtime) {
					let res = {};
					if ((event.data.corsProxy.webRequests != undefined) || !!webRequests) {
						res.webRequests = webRequests;
						if ((event.data.corsProxy.webRequests == true) && !webRequests)
							webRequests = [];
						else if ((event.data.corsProxy.webRequests == false) && !!webRequests)
							webRequests = null;
					}
					if (event.data.corsProxy.eval) {
						res.eval = new Function('return ('+event.data.corsProxy.eval+')()')();
						if (res.eval && (res.eval.constructor.name == 'Promise'))
							res.eval = await res.eval;
					}
					event.source.postMessage({
						corsProxy: res,
						timeStamp: event.data.timeStamp,
						runtime: true
					}, '*');
				}
			});
		}
	});
}


extension/background.js
let config;
let headerFind = (headers, key, value) => {
	let o = headers.find(({ name }) => name.match(new RegExp(key, 'i')));
	return (o ? o.value : (value || null));
}
let proxy = () => chrome.proxy.settings.set({
	value: (!config.antizapret ? { mode: 'system' } : {
		mode: 'pac_script',
		pacScript: {
			url: 'https://antizapret.prostovpn.org/proxy.pac',
			mandatory: true
		}
	}),
	scope: 'regular'
});
let menu = {
	cors: ps => ({
		id: 'cors',
		title: 'Enable cors',
		type: 'checkbox',
		checked: ps.enabled,
		contexts: [ 'browser_action' ],
		onclick: ({ checked, update }) => {
			ps.enabled = checked;
			chrome.storage.local.set(ps);
			if (!ps || update)
				chrome.contextMenus.update('cors', { checked })
		}
	}),
	antizapret: ps => ({
		id: 'antizapret',
		title: 'Antizapret',
		type: 'checkbox',
		checked: (ps || config).antizapret,
		contexts: [ 'browser_action' ],
		onclick: ({ checked, update }) => {
			proxy((ps || config).antizapret = checked);
			chrome.storage.local.set(ps || config);
			if (!ps || update)
				chrome.contextMenus.update('antizapret', { checked })
		}
	}),
	cookies: ps => ({
		title: 'Save cookies',
		contexts: [ 'browser_action' ],
		onclick: () => chrome.tabs.getSelected(null, tab => chrome.cookies.getAll({
			url: (tab.url.replace(/^chrome.*:\/\/.*/, '') || null)
		}, cookies => chrome.downloads.download({
			filename: 'cookies.txt',
			url: URL.createObjectURL(new Blob([
				[
					'# This file can be used by wget, curl, aria2c and other standard compliant tools.',
					'# Usage Examples:',
					'#   1) wget -x --load-cookies cookies.txt "'+tab.url+'"',
					'#   2) curl --cookie cookies.txt "'+tab.url+'"',
					'#   3) aria2c --load-cookies cookies.txt "'+tab.url+'"',
					''
				].concat(cookies.map(v => [
					v.domain,
					(v.hostOnly ? 'FALSE' : 'TRUE'),
					v.path,
					(v.secure ? 'TRUE' : 'FALSE'),
					(v.expirationDate ? Math.round(v.expirationDate) : '0'),
					v.name,
					v.value
				].join('\t'))).join('\n')
			], { type: 'text/plain;charset=utf-8' }))
		})))
	})
};
chrome.storage.local.get({
	enabled: true,
	antizapret: false
	// puppeteer: null
}, /* async */ ps => {
	config = ps;
	Object.entries(chrome.runtime.getManifest().config || {}).forEach(e => (ps[e[0]] = e[1])); // fetch('chrome://version/strings.js')
	if (!location.href.endsWith('background.html')) {
		if (ps.antizapret)
			proxy();
		/*
		ps.puppeteer = await fetch("http://localhost:9222/json/version").then(res => res.json()).then(data => data.webSocketDebuggerUrl).catch(err => null);
		chrome.storage.local.set(ps);
		chrome.contextMenus.create(
			ps.puppeteer ? {
				title: 'Puppeteer',
				type: 'checkbox',
				checked: true,
				enabled: false,
				contexts: [ 'browser_action' ]
			} : {
				title: 'Puppeteer',
				enabled: true,
				contexts: [ 'browser_action' ],
				onclick: () => alert('Add --remote-debugging-port=9222 to the end of the target field of your chrome.exe short cut.')
			}
		);
		*/
		Object.values(menu).forEach(m => chrome.contextMenus.create(m(ps)));
		chrome.webRequest.onBeforeSendHeaders.addListener(({ url, requestHeaders }) => (!ps.enabled ? {} : {
			requestHeaders: requestHeaders.map(h => {
				let i, k;
				if (h.name.match(/^cors-/i) && (k = h.name.replace(/^cors-/i, ''))) {
					if ((i = requestHeaders.findIndex(h_ => h_.name.match(new RegExp('^'+k+'$', 'i')))) > -1)
						requestHeaders[i].value = null;
					return { name: k, value: ((h.value == 'null') ? h.value : null) };
				}
				return h;
			})
			// .filter(h => (h.value && (h.value != 'null')))
			.filter(Boolean)
			.concat(Object.entries(
				JSON.parse(Object.fromEntries(new URLSearchParams('?'+url.split('?')[1])).corsProxy || '{}')
			).map(h => ({
				name: h[0], value: h[1]
			})))
		}), {
			urls: [	'<all_urls>' ]
		}, [
			'blocking', 'requestHeaders', 'extraHeaders'
		]);
		chrome.webRequest.onHeadersReceived.addListener(e => {
			chrome.tabs.sendMessage(e.tabId, { webRequest: e }); 
			let responseHeaders = e.responseHeaders;
			return (!ps.enabled ? {} : {
				responseHeaders: responseHeaders.concat([
					{ name: 'Access-Control-Allow-Origin', value: headerFind(responseHeaders, 'access-control-allow-origin', '*') },
					{ name: 'Access-Control-Allow-Methods', value: headerFind(responseHeaders, 'access-control-allow-methods', '*') },
					{ name: 'Access-Control-Allow-Headers', value: headerFind(responseHeaders, 'access-control-allow-headers', '*') },
					{ name: 'Access-Control-Expose-Headers', value: '*' },
					{ name: 'Access-Control-Allow-Credentials', value: 'true' }
				])
			});
		}, {
			urls: [	'<all_urls>' ]
		}, [
			'blocking', 'responseHeaders'
		]);
		chrome.browserAction.onClicked.addListener(() => {
			chrome.windows.create({
				url: chrome.runtime.getURL('background.html'),
				type: 'popup',
				height: 330,
				width: 250
			})
		});
	}else{
		let content = document.querySelector('main section');
		Object.values(menu).map(m => m(ps)).forEach(m => {
			let tpl = document.createElement('template');
			if (m.type == 'checkbox')
				tpl.innerHTML = '<label class="checkbox"><input type="checkbox" '+(m.checked ? 'checked' : '')+'><span>'+m.title+'</span></label>';
			else
				tpl.innerHTML = '<label class="button"><input type="button" value="'+m.title+'"/></label>';
			tpl.content.querySelector('input').addEventListener('click', event => m.onclick({
				update: true,
				checked: event.target.checked
			}));
			content.append(tpl.content);
		});
	}
});


worker.js
self.addEventListener('fetch', async event => {
	const
		req = event.request,
		url = new URL(req.url);
	if ((url.hostname == 'localhost') && (url.pathname != '/iscorsneeded'))
		return;
	event.respondWith((async () => {
		let res;
		try {
			if (url.pathname === '/')
				res = new Response(url.origin+'/<url>');
			else if (url.pathname === '/iscorsneeded')
				return new Response((req.headers.get('sec-fetch-dest')) ? 'yes' : 'no');
			else if ((!self.location || (self.location.hostname == url.hostname)) && !url.pathname.startsWith('/http'))
				res = new Response('Not Found', { 
					status: 404,
					statusText: 'Not Found',
					headers: { 'Content-Type': 'text/html' }
				});
			else{
				if (req.method == 'OPTIONS')
					res = new Response(null, {
						status: 200,
						statusText: 'OK'
					});
				else{
					let headers = {};
					if (self.location)
						headers = req.headers;
					else{
						for (let h of req.headers.entries()) {
							if (!h[0].match(/^origin/i) && !h[0].match(/^cf-/i) && !h[0].match(/^x-forw/i) && !h[0].match(/^cors-/i))
								headers[h[0]] = h[1];
						}
						for (let h of req.headers.entries()) {
							if (h[0].match(/^cors-/i) && ([ 'cache' /* , 'redirect' */ ].indexOf(h[0] = h[0].replace(/^cors-/i, '')) == -1)) {
								if (h[1] && (h[1] != 'null'))
									headers[h[0]] = h[1];
								else
									delete headers[h[0]];
							}
						}
                        let hh = JSON.parse(Object.fromEntries(new URLSearchParams(url.search)).corsProxy || '{}');
                        for (let h in hh) {
                            headers[h] = hh[h];
                        }
					}
					let get = await fetch((url.pathname.startsWith('/http') ? req.url.slice(url.origin.length + 1) : url.href), {
						method: req.method,
						headers: headers,
						redirect: 'follow', // (req.headers.get('Cors-Redirect') || 'follow'),
						body: req.body
					});
					res = new Response(get.body, get);
				}
			}
			if (req.headers.get('Origin')) {
				res.headers.set('Access-Control-Allow-Origin', req.headers.get('Origin'));
				res.headers.set('Access-Control-Allow-Credentials', 'true');
				if (req.headers.get('Access-control-request-headers'))
					res.headers.set('Access-Control-Allow-Headers', req.headers.get('Access-control-request-headers'));
				res.headers.set('Access-Control-Expose-Headers', '*');
			}
			if (req.headers.get('Cors-Cache')) {
				res.headers.set('Cache-control', 'public');
				res.headers.set('Expires', new Date(Date.now() + parseInt(req.headers.get('Cors-Cache'))).toUTCString());
			}
			// if (req.headers.get('Cors-Redirect'))
				// res.headers.delete('Location');
		} catch (e) {
			res = new Response((e.stack || e), { status: 500 });
		}
		return res;
	})());
});
